File: solutions/typescript/binary-search/1/binary-search.ts
export function find<T>(haystack: Array<T>, needle: T): number | never {
  let low = 0
  let high = haystack.length

  while (low < high) {
    const mid = Math.trunc((low + high) / 2)
    const value = haystack[mid]

    if (value === needle) return mid

    if (value < needle) {
      low = mid + 1
    } else {
      high = mid
    }
  }

  throw new Error('Value not in array')
}

File: solutions/go/binary-search/1/binary_search.go
package binarysearch

func SearchInts(list []int, key int) int {
	low := 0
	high := len(list)

	for low < high {
		mid := (low + high) / 2
		value := list[mid]

		if value == key {
			return mid
		}

		if value < key {
			low = mid + 1
		} else {
			high = mid
		}
	}

	return -1
}
